### PR TITLE
Add permission for reopening terminal/closed cases

### PR DESCRIPTION
**Pull request by stephanieboyd**

### CODE REVIEW

File: applications/libraries/application_helpers.py [hunk 1]
```diff
@@ -25,6 +27,20 @@ def validate_status_can_be_set_by_exporter_user(original_status: str, new_status
...
-def validate_status_can_be_set_by_gov_user(status: str) -> Optional[str]:
-    if status == CaseStatusEnum.APPLICANT_EDITING:
-        return f'Setting application status to "{status}" is not allowed for GovUsers.'
+def validate_status_can_be_set_by_gov_user(user, original_status: str, new_status: str) -> Optional[str]:
+    """ Validate that a status can be set by a gov user. Gov users can not set a case's status to
+    Applicant editing. They also cannot set a case's status to `Finalised` or open a closed case
+    without additional permissions.
+    """
+    if new_status == CaseStatusEnum.APPLICANT_EDITING:
+        return f'Setting application status to "{new_status}" is not allowed for GovUsers.'
```
Comment (jfaracik) [inline applications/libraries/application_helpers.py]: Shouldd this return a string, or should it raise an error?

Comment (jfaracik) [reply applications/libraries/application_helpers.py]: Raising a validation error would be better, like so: `raise serializers.ValidationError({"errors": {"sites": ["Select at least one site to assign the user to"]}})`

File: applications/tests/tests_application_status.py [hunk 2]
```diff
@@ -143,3 +143,34 @@ def test_gov_set_status_for_all_except_applicant_editing_and_finalised_success(s
...
+        self.standard_application.refresh_from_db()
+
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
+        self.assertEqual(self.standard_application.status, get_case_status_by_status(CaseStatusEnum.WITHDRAWN))
+
+    def test_gov_set_status_when_they_have_permission_to_reopen_closed_cases_success(self):
+        self.standard_application.status = get_case_status_by_status(CaseStatusEnum.WITHDRAWN)
+        self.standard_application.save()
+
+        # Give gov user super used role, to include reopen closed cases permission
```
Comment (jfaracik) [inline applications/tests/tests_application_status.py]: *Super **user**

Comment (stephanieboyd) [reply applications/tests/tests_application_status.py]: Good reading, thanks!

File: applications/views/applications.py [hunk 3]
```diff
@@ -214,24 +213,23 @@ def put(self, request, pk):
...
         else:
-            validation_error = validate_status_can_be_set_by_gov_user(new_status_enum)
-
-        if validation_error:
-            return JsonResponse(data={"errors": [validation_error]}, status=status.HTTP_400_BAD_REQUEST)
-
-        # Only allow status change to FINALISED if user has MANAGE_FINAL_ADVICE permission
-        if new_status_enum == CaseStatusEnum.FINALISED:
-            assert_user_has_permission(request.user, constants.GovPermissions.MANAGE_FINAL_ADVICE)
+            if not check_status_can_be_set_by_gov_user(request.user, application.status.status, new_status):
```
Comment (andytaylorkainos) [inline applications/views/applications.py]: same as above

File: queries/end_user_advisories/tests/test_create_end_user_advisories.py [hunk 4]
```diff
@@ -288,23 +232,4 @@ def test_create_end_user_advisory_query_for_individual(self):
...
-    def test_update_end_user_advisory_status_success(self):
-        data = {"status": CaseStatusEnum.RESUBMITTED}
-
-        response = self.client.put(self.url, data, **self.gov_headers)
-        self.assertEqual(response.status_code, status.HTTP_200_OK)
-
-        new_end_user_advisory = EndUserAdvisoryQuery.objects.get(pk=self.end_user_advisory.id)
-        case_status = get_case_status_by_status(CaseStatusEnum.RESUBMITTED)
-        self.assertEqual(new_end_user_advisory.status, case_status)
+        self.assertEqual(Case.objects.count(), 1)
```
Comment (andytaylorkainos) [inline queries/end_user_advisories/tests/test_create_end_user_advisories.py]: new line

File: queries/end_user_advisories/tests/test_end_user_advisories_status.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        self.query.save()
+
+        self.url = reverse("queries:end_user_advisories:end_user_advisory", kwargs={"pk": self.query.id})
+
+    def test_gov_set_status_when_no_permission_to_reopen_closed_cases_failure(self):
+        data = {"status": CaseStatusEnum.SUBMITTED}
+
+        response = self.client.put(self.url, data=data, **self.gov_headers)
+
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
```
Comment (andytaylorkainos) [inline queries/end_user_advisories/tests/test_end_user_advisories_status.py]: can you assert that the status wasnt changed please

File: queries/end_user_advisories/tests/test_end_user_advisories_status.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        data = {"status": CaseStatusEnum.SUBMITTED}
+
+        # Give gov user super used role, to include reopen closed cases permission
+        self.gov_user.role = self.super_user_role
+        self.gov_user.save()
+
+        response = self.client.put(self.url, data=data, **self.gov_headers)
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
```
Comment (andytaylorkainos) [inline queries/end_user_advisories/tests/test_end_user_advisories_status.py]: can you assert the new status please

File: queries/end_user_advisories/tests/test_update_end_user_advisories.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+    def test_update_end_user_advisory_status_success(self):
+        data = {"status": CaseStatusEnum.RESUBMITTED}
+
+        response = self.client.put(self.url, data, **self.gov_headers)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+        new_end_user_advisory = EndUserAdvisoryQuery.objects.get(pk=self.end_user_advisory.id)
+        case_status = get_case_status_by_status(CaseStatusEnum.RESUBMITTED)
+        self.assertEqual(new_end_user_advisory.status, case_status)
```
Comment (andytaylorkainos) [inline queries/end_user_advisories/tests/test_update_end_user_advisories.py]: new line

File: queries/end_user_advisories/tests/test_view_end_user_advisories.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
+from rest_framework import status
+from rest_framework.reverse import reverse
+
+from test_helpers.clients import DataTestClient
+
+
+class EndUserAdvisoryViewTests(DataTestClient):
+    def test_view_end_user_advisory_queries(self):
```
Comment (andytaylorkainos) [inline queries/end_user_advisories/tests/test_view_end_user_advisories.py]: _success

File: queries/end_user_advisories/tests/test_view_end_user_advisories.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+        end_user_data = response_data["end_user"]
+        self.assertEqual(end_user_data["sub_type"]["key"], query.end_user.sub_type)
+
+        self.assertEqual(end_user_data["name"], query.end_user.name)
+        self.assertEqual(end_user_data["website"], query.end_user.website)
+        self.assertEqual(end_user_data["address"], query.end_user.address)
+        self.assertEqual(end_user_data["country"]["id"], query.end_user.country.id)
+
+    def test_view_end_user_advisory_query_on_organisation(self):
```
Comment (andytaylorkainos) [inline queries/end_user_advisories/tests/test_view_end_user_advisories.py]: _success

File: queries/end_user_advisories/tests/test_end_user_advisories_status.py [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    def test_gov_set_status_when_they_have_permission_to_reopen_closed_cases_success(self):
+        data = {"status": CaseStatusEnum.SUBMITTED}
+
+        # Give gov user super used role, to include reopen closed cases permission
+        self.gov_user.role = self.super_user_role
+        self.gov_user.save()
+
+        response = self.client.put(self.url, data=data, **self.gov_headers)
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
```
Comment (andytaylorkainos) [inline queries/end_user_advisories/tests/test_end_user_advisories_status.py]: can you assert that the status was changed please

File: queries/end_user_advisories/tests/test_view_end_user_advisories.py (hunk 9 above)
Comment (stephanieboyd) [reply queries/end_user_advisories/tests/test_view_end_user_advisories.py]: I just moved these into their own files, won't be changing :)

File: queries/end_user_advisories/tests/test_view_end_user_advisories.py (hunk 8 above)
Comment (stephanieboyd) [reply queries/end_user_advisories/tests/test_view_end_user_advisories.py]: Just moved these tests into their own file, won't be changing :)

Comment (stephanieboyd) [reply queries/end_user_advisories/tests/test_view_end_user_advisories.py]: I just moved these, won't be changing :)

File: applications/libraries/application_helpers.py [hunk 11]
```diff
@@ -14,17 +14,36 @@ def optional_str_to_bool(optional_string: str):
         raise ValueError("You provided " + optional_string + ', while the allowed values are None, "true" or "false"')
 
 
-def validate_status_can_be_set_by_exporter_user(original_status: str, new_status: str) -> Optional[str]:
-    error_message_template = 'Setting application status to "{}" when application status "{}" is not allowed.'
-
+def check_status_can_be_set_by_exporter_user(original_status: str, new_status: str) -> bool:
+    """ Check that a status can be set by an exporter user. Exporter users cannot withdrawn an application
```
[the remark this answers is not in the document]
Comment (stephanieboyd) [reply applications/libraries/application_helpers.py]: Fixed

File: applications/libraries/application_helpers.py [hunk 12]
```diff
@@ -14,17 +14,36 @@ def optional_str_to_bool(optional_string: str):
...
     elif CaseStatusEnum.is_read_only(original_status) or new_status != CaseStatusEnum.APPLICANT_EDITING:
-        return error_message_template.format(new_status, original_status)
+        return False
+
+    return True
+
 
+def check_status_can_be_set_by_gov_user(user, original_status: str, new_status: str) -> bool:
+    """ Check that a status can be set by a gov user. Gov users can not set a case's status to
+    Applicant editing. They also cannot set a case's status to `Finalised` or open a closed case
```
[the remark this answers is not in the document]
Comment (stephanieboyd) [reply applications/libraries/application_helpers.py]: Wrapped applicant editing in quotes

File: applications/tests/tests_application_status.py [hunk 13]
```diff
@@ -25,8 +25,7 @@ def test_gov_set_application_status_to_applicant_editing_failure(self):
         self.standard_application.refresh_from_db()
         self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
         self.assertEqual(
-            json.loads(response.content).get("errors")[0],
-            'Setting application status to "applicant_editing" is not allowed for GovUsers.',
+            json.loads(response.content).get("errors")[0], "Status cannot be set by Gov user.",
```
[the remark this answers is not in the document]
Comment (stephanieboyd) [reply applications/tests/tests_application_status.py]: None, it was just reworked from what previously existed. Made your suggested change

File: applications/views/applications.py [hunk 14]
```diff
@@ -214,24 +213,23 @@ def put(self, request, pk):
...
         data = request.data
-        new_status_enum = data.get("status")
+        new_status = data.get("status")
 
         if isinstance(request.user, ExporterUser):
             if request.user.organisation.id != application.organisation.id:
                 raise PermissionDenied()
 
-            validation_error = validate_status_can_be_set_by_exporter_user(application.status.status, new_status_enum)
+            if not check_status_can_be_set_by_exporter_user(application.status.status, new_status):
```
[the remark this answers is not in the document]
Comment (stephanieboyd) [reply applications/views/applications.py]: I've changed the function name to `can` but I like the `if not` and bailing out check

File: applications/views/applications.py (hunk 3 above)
Comment (stephanieboyd) [reply applications/views/applications.py]: Same as above

File: queries/end_user_advisories/tests/test_end_user_advisories_status.py (hunk 5 above)
Comment (stephanieboyd) [reply queries/end_user_advisories/tests/test_end_user_advisories_status.py]: yeah, good spot, i do in the other tests - dunno what happened here

File: queries/end_user_advisories/tests/test_end_user_advisories_status.py (hunk 10 above)
Comment (stephanieboyd) [reply queries/end_user_advisories/tests/test_end_user_advisories_status.py]: Same as above

File: applications/libraries/application_helpers.py [hunk 15]
```diff
@@ -14,17 +14,36 @@ def optional_str_to_bool(optional_string: str):
         raise ValueError("You provided " + optional_string + ', while the allowed values are None, "true" or "false"')
 
 
-def validate_status_can_be_set_by_exporter_user(original_status: str, new_status: str) -> Optional[str]:
-    error_message_template = 'Setting application status to "{}" when application status "{}" is not allowed.'
-
+def can_status_can_be_set_by_exporter_user(original_status: str, new_status: str) -> bool:
+    """ Check that a status can be set by an exporter user. Exporter users cannot withdraw an application
+    that is already in a terminal state and they cannot set an application to applicant editing if the
```
Comment (andytaylorkainos) [inline applications/libraries/application_helpers.py]: `Applicant editing` - approving, but if you could just do this it'd be awesome :)

File: queries/end_user_advisories/tests/test_end_user_advisories_status.py (hunk 6 above)
Comment (stephanieboyd) [reply queries/end_user_advisories/tests/test_end_user_advisories_status.py]: done

**Pull request merged**